File: Linear_Control_Course/Chapter15/Lesson3/Chapter15_Lesson3.java
import org.apache.commons.math3.complex.Complex;

public class NyquistRelativeStability {

    // Example: G(s) = 1 / (s (s + 1) (s + 2)), L(s) = K G(s)
    public static Complex Lofjw(double omega, double K) {
        Complex j = new Complex(0.0, 1.0);
        Complex s = j.multiply(omega);
        Complex denom = s.multiply(s.add(1.0)).multiply(s.add(2.0));
        return new Complex(K, 0.0).divide(denom);
    }

    public static void main(String[] args) {
        double K = 2.0;
        double dMin = Double.POSITIVE_INFINITY;
        double omegaMin = 0.0;

        for (int k = 0; k <= 1000; ++k) {
            double omega = Math.pow(10.0, -2.0 + 4.0 * k / 1000.0);
            Complex L = Lofjw(omega, K);
            Complex z = L.add(1.0); // 1 + L(jw)
            double d = z.abs();
            if (d < dMin) {
                dMin = d;
                omegaMin = omega;
            }
        }

        System.out.println("Approx d_min = " + dMin + " at omega = " + omegaMin);

        // In a robotics codebase, this can be used to check stability of joint controllers
        // when gains are auto-tuned or adapted online.
    }
}

File: Tutorials/Codes/Linear_Control_Course/Chapter15/Lesson3/Chapter15_Lesson3.cpp
#include <iostream>
#include <complex>
#include <vector>
#include <cmath>

// Example: G(s) = 1 / (s (s + 1) (s + 2)), L(s) = K G(s)
std::complex<double> L_of_jw(double omega, double K) {
    std::complex<double> j(0.0, 1.0);
    std::complex<double> s = j * omega;
    std::complex<double> denom = s * (s + 1.0) * (s + 2.0);
    return K / denom;
}

int main() {
    double K = 2.0;
    std::vector<double> omega_vec;
    for (int k = 0; k <= 1000; ++k) {
        double omega = std::pow(10.0, -2.0 + 4.0 * k / 1000.0);
        omega_vec.push_back(omega);
    }

    double d_min = 1e9;
    double omega_min = 0.0;

    for (double omega : omega_vec) {
        std::complex<double> L = L_of_jw(omega, K);
        std::complex<double> z = 1.0 + L; // point in (1 + L(jw))-plane
        double d = std::abs(z);
        if (d < d_min) {
            d_min = d;
            omega_min = omega;
        }
    }

    std::cout << "Approx d_min = " << d_min
              << " at omega = " << omega_min << std::endl;

    // This data can be published to ROS topics or logged for plotting in Python/Matlab.
    return 0;
}
